src/layout.cpp

```diff
@@ -15,84 +15,62 @@
-    return !tensor.empty() || !model.empty();
-}
-
-Status LayoutConfiguration::fromString(const std::string& configuration, LayoutConfiguration& configOut) {
-    std::string configurationCopy = configuration;
-    erase_spaces(configurationCopy);
-
-    std::transform(configurationCopy.begin(), configurationCopy.end(), configurationCopy.begin(), ::toupper);
-
-    if (configurationCopy.find_first_not_of("NCHWD?.:") != std::string::npos)
-        return StatusCode::LAYOUT_WRONG_FORMAT;
-
-    size_t delimCount = std::count(configurationCopy.begin(), configurationCopy.end(), ':');
-    if (delimCount > 1)
-        return StatusCode::LAYOUT_WRONG_FORMAT;
-
-    if (delimCount == 0) {
-        configOut = LayoutConfiguration(configurationCopy);
-    } else {
-        std::vector<std::string> tokens = tokenize(configurationCopy, ':');
-        if (tokens.size() > 2)
-            return StatusCode::LAYOUT_WRONG_FORMAT;
-        else if (tokens.size() == 2)
-            configOut = LayoutConfiguration(tokens[0], tokens[1]);
-        else if (tokens.size() == 1)
-            configOut = LayoutConfiguration(tokens[0]);
-        else
-            return StatusCode::LAYOUT_WRONG_FORMAT;
+std::optional<size_t> Layout::retrieveBatchIndex() const {
+    auto status = validate();
+    if (!status.ok()) {
+        return std::nullopt;
     }
-    return StatusCode::OK;
-}
-
-std::string LayoutConfiguration::toString() const {
-    std::stringstream ss;
-    if (tensor.empty()) {
-        ss << model;
-    } else {
-        ss << tensor << ":" << model;
+    auto batchPos = this->find('N');
+    auto etcPos = this->find("...");
+    if (batchPos == std::string::npos) {
+        return std::nullopt;
+    }
+    if (etcPos != std::string::npos && batchPos > etcPos) {
+        return std::nullopt;
     }
-    return ss.str();
+    return batchPos;
 }
 
-const std::string& LayoutConfiguration::getTensorLayout() const {
-    return this->tensor;
-}
+Status Layout::validate() const {
+    if (this->find_first_not_of("NCHWD?.") != std::string::npos)
+        return StatusCode::LAYOUT_WRONG_FORMAT;  // Cannot contain other letters
 
-const std::string& LayoutConfiguration::getModelLayout() const {
-    return this->model;
-}
+    for (char c : std::string{"NCHWD"}) {
+        if (std::count(this->begin(), this->end(), c) > 1) {
+            return StatusCode::LAYOUT_WRONG_FORMAT;  // Can contain NCHWD only single time
+        }
+    }
 
-bool LayoutConfiguration::operator==(const LayoutConfiguration& rhs) const {
-    return this->tensor == rhs.tensor && this->model == rhs.model;
-}
+    size_t dotCount = 0;
+    bool etcAppeared = false;
+    for (char c : (*this)) {
+        if (c == '.') {
+            if (etcAppeared) {
+                return StatusCode::LAYOUT_WRONG_FORMAT;  // Cannot appear multiple times
+            }
+            dotCount++;
+            if (dotCount >= 3) {
+                etcAppeared = true;
+            }
+        }
+    }
```

Comment:
WHat about handling eg NHW../NH.? Shouldn't those be illegal?
The same about N.H.W

Reply:
done